### DISCUSSION TITLE
Results on Local Nominatim Different Than OpenStreetMap.org

### DISCUSSION THREAD

**Discussion by rmmays** [osm-search/Nominatim#2176] [Q&A]
Hi Everyone,
I have deployed a local install of Nominatim.  Loading the globe, wiki, tiger data all went well with no errors.  However, when I search an address (e.g. 4922 N Harbor Dr, San Diego, CA 92106) it shows up "close" to the correct spot, but the openstreetmap.org site shows _exactly_ the right spot.  I am wondering if there's something I can do to reload my data or re-index it for better results?  I am wondering if it is ignoring the house address (4922).  It seems to just be finding the street name and showing where it starts.

Correct Result on openstreetmaps.org
<img width="1680" alt="Screen Shot 2021-02-17 at 12 00 56 AM" src="https://user-images.githubusercontent.com/1320049/108162708-45f3e200-70b3-11eb-9756-4a1455b15246.png">

My Local Deployment "Incorrect" Result
<img width="1680" alt="Screen Shot 2021-02-16 at 11 53 38 PM" src="https://user-images.githubusercontent.com/1320049/108162374-9880ce80-70b2-11eb-8ed4-e5b34a9f8f1a.png">

My Local Deployment "See More" for Incorrect Result
<img width="1680" alt="Screen Shot 2021-02-16 at 11 55 56 PM" src="https://user-images.githubusercontent.com/1320049/108162369-96b70b00-70b2-11eb-96dd-6ef0faa25925.png">

**Comment by mtmail** [comment]
That looks like TIGER data isn't imported or queried.

* Does the debug output (e.g. https://nominatim.openstreetmap.org/search.php?q=4922%20N%20Harbor%20Dr%2C%20San%20Diego%2C%20CA%2092106&polygon_geojson=1&format=jsonv2&debug=1) show `location_property_tiger` being queried?

* Can you check the number of rows in the `location_property_tiger` table? Should be about 30 million.

* Do you have `@define('CONST_Use_US_Tiger_Data', true);` enabled in `build/settings/local.php` (might require restart of the webserver).

**Reply by rmmays** [reply]
@mtmail so grateful for your reply!!! Thank you!
1. the debug output does not contain any instances of location_property_tiger (i used Ctrl + F to search for it in results)

I have attached the full output from the debug query you gave me: [debug_output.txt](https://github.com/osm-search/Nominatim/files/5996900/debug_output.txt)
Here it is in .htm format (remove .txt in filename): [debug_output.htm.txt](https://github.com/osm-search/Nominatim/files/5996934/debug_output.htm.txt)

2. there are 33,496,424 entries in my location_property_tiger table

3. I have the following in my build/settings/local.php (and I have the beginning php declaration on first line, but this text box won't let me put it in my reply here....):

@define('CONST_Pyosmium_Binary', '/home/nominatim/.local/bin/pyosmium-get-changes');
@define('CONST_Osm2pgsql_Flatnode_File', '/home/nominatim/Nominatim-3.6.0/data/flatnode.file');
@define('CONST_Debug', false);
@define('CONST_Search_BatchMode', true);
@define('CONST_BasePath', '/home/nominatim/Nominatim-3.6.0');
@define('CONST_Database_Web_User', 'apache');
@define('CONST_Import_Style', CONST_BasePath.'/settings/import-extratags.style');
@define('CONST_Use_US_Tiger_Data', true);

local.php file: [local.php.txt](https://github.com/osm-search/Nominatim/files/5996910/local.php.txt)

**Reply by mtmail** [reply]
Configuration looks alright. Please double-check that the webserver (what executes PHP, could also be php-fpm in your setup) is restarted to load the new configuration.

If that fails you need a bit of debugging. In `lib/SearchDescription.php` you will see a line `if (CONST_Use_US_Tiger_Data && $bIsIntHouseNumber && empty($aResults)) {` Based on your debug output house number is set and results are empty, so question remains if the first value is true.

You can add a statement like
`echo "hello"; echo var_dump(name-of-variable-or-data-structure);`
which will show up on the &debug=1 output.

Further below you'll see `Debug::printSQL($sSQL);`. `lib/DebugHTML.php` has function for pretty tables, headers, essentially they execute `echo` though.

**Reply by rmmays** [reply]
@mtmail, your response time is amazing!!!!
I will try your suggestions above and let you know what I find.  Have a nice night in Germany.  Hopefully I will have a few answers for you by your morning.

**Reply by rmmays** [reply]
@mtmail, I put in the debug line you suggested.  it looks like the CONST_Use_US_Tiger_Data doesn't have any value...
I put in:

// If nothing found then search in Tiger data (location_property_tiger)
echo "TigerDebug: CONST_Use_US_Tiger_Data: " . CONST_Use_US_Tiger_Data . ", bIsIntHouseNumber: " . $bIsIntHouseNumber . ", empty(aResults):" . empty($aResults);
if (CONST_Use_US_Tiger_Data && $bIsIntHouseNumber && empty($aResults)) {
 ....
}

My debug output now says:
TigerDebug: CONST_Use_US_Tiger_Data: , bIsIntHouseNumber: 1, empty(aResults):1

So, it seems like my @define('CONST_Use_US_Tiger_Data', true); isn't getting picked up out of my local.php file.
i have this file at build/settings/local.php.    Is that correct?  Is there a way to debug whether my local settings are getting picked up?

**Reply by mtmail** [reply]
Can you check `build/settings/setting-frontend.php`?

It gets created by `utils/setup.php --setup-website` (see https://nominatim.org/release-docs/3.6.0/appendix/Install-on-Ubuntu-20/#setting-up-a-webserver) and takes values from the existing configuration to create this new configuration file, code is at https://github.com/osm-search/Nominatim/blob/v3.6.0/lib/setup/SetupClass.php#L705 It's possible if you had `CONST_Use_US_Tiger_Data` set to false during the initial setup then the value in `settings-frontend.php` is still false.

**Reply by rmmays** [reply]
i'm not sure i know how to properly print out the CONST_ variables.  I used PHP like 15 years ago and haven't touched it since, so I have no idea what I'm doing.... :-)

// If nothing found then search in Tiger data (location_property_tiger)
echo "TigerDebug: CONST_Use_US_Tiger_Data: " . CONST_Use_US_Tiger_Data . ", bIsIntHouseNumber: " . $bIsIntHouseNumber . ", empty(aResults):" . empty($aResults) . "<br>";
echo "CONST_Pyosmium_Binary: " . constant(CONST_Pyosmium_Binary) . "<br>";
echo "CONST_Osm2pgsql_Flatnode_File: " . constant(CONST_Osm2pgsql_Flatnode_File) . "<br>";
echo "CONST_Debug: " . CONST_Debug . "<br>";
echo "CONST_Search_BatchMode: " . CONST_Search_BatchMode . "<br>";
echo "CONST_BasePath: " . CONST_BasePath . "<br>";
echo "CONST_Database_Web_User: " . CONST_Database_Web_User . "<br>";
echo "CONST_Import_Style: " . CONST_Import_Style . "<br>";
echo "CONST_Use_US_Tiger_Data: " . CONST_Use_US_Tiger_Data . "<br>";

Produces......

TigerDebug: CONST_Use_US_Tiger_Data: , bIsIntHouseNumber: 1, empty(aResults):1
CONST_Pyosmium_Binary:
CONST_Osm2pgsql_Flatnode_File:
CONST_Debug: 1
CONST_Search_BatchMode: 1
CONST_BasePath: /home/nominatim/Nominatim-3.6.0
CONST_Database_Web_User: CONST_Database_Web_User
CONST_Import_Style: CONST_Import_Style
CONST_Use_US_Tiger_Data:

In my debug output

**Reply by rmmays** [reply]
OOOH!!!!!  In my settings-frontend.php is is FALSE!!!!

@define('CONST_BasePath', '/home/nominatim/Nominatim-3.6.0');
if (file_exists(getenv('NOMINATIM_SETTINGS'))) require_once(getenv('NOMINATIM_SETTINGS'));

@define('CONST_Database_DSN', 'pgsql:dbname=nominatim');
@define('CONST_Default_Language', false);
@define('CONST_Log_DB', false);
@define('CONST_Log_File', false);
@define('CONST_Max_Word_Frequency', '50000');
@define('CONST_NoAccessControl', 1);
@define('CONST_Places_Max_ID_count', 50);
@define('CONST_PolygonOutput_MaximumTypes', 1);
@define('CONST_Search_AreaPolygons', 1);
@define('CONST_Search_BatchMode', true);
@define('CONST_Search_NameOnlySearchFrequencyThreshold', 500);
@define('CONST_Search_ReversePlanForAll', 1);
@define('CONST_Term_Normalization_Rules', ":: NFD (); [[:Nonspacing Mark:] [:Cf:]] >;  :: lower (); [[:Punctuation:][:Space:]]+ > ' '; :: NFC ();");
@define('CONST_Use_Aux_Location_data', false);
@define('CONST_Use_US_Tiger_Data', false);
@define('CONST_MapIcon_URL', false);

**Reply by rmmays** [reply]
I updated to true in my settings-frontend.php and that was it!!!!  It works now and shows the properly placed spot on the address!!!

Now, do I need to re-run some of the setup stuff to make sure I don't have something else messed up?  I tried to follow all the directions, but it seems like some of the procedures are for older versions.  I don't know how to get GB tiger data in and i'm not sure if the wiki data worked either....

if i want ALL available data in there, what do i need to download and run?

**Reply by mtmail** [reply]
Great. US TIGER will work now, no further changes needed.

For wikipedia/wikidata see https://nominatim.org/release-docs/3.6.0/admin/Import/#wikipediawikidata-rankings There is a "If you forgot to download" section.

For US or GB postcode data you can download
```
cd $NOMINATIM_SOURCE_DIR/data
wget https://www.nominatim.org/data/gb_postcode_data.sql.gz
wget https://www.nominatim.org/data/us_postcode_data.sql.gz
```
and pipe the data into the database, e.g. `zcat $file | pql nominatim` to fill the database tables. After run `build/utils/update.php --calculate-postcodes`, which processes the data into the `location_postcodes` table.
